controller/pkg/aclprovider/iptablesprovider.go

```diff
@@ -0,0 +1,299 @@
+package provider
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/coreos/go-iptables/iptables"
+	"go.uber.org/zap"
+)
+
+// IptablesProvider is an abstraction of all the methods an implementation of userspace
+// iptables need to provide.
+type IptablesProvider interface {
+	// Append apends a rule to chain of table
+	Append(table, chain string, rulespec ...string) error
+	// Insert inserts a rule to a chain of table at the required pos
+	Insert(table, chain string, pos int, rulespec ...string) error
+	// Delete deletes a rule of a chain in the given table
+	Delete(table, chain string, rulespec ...string) error
+	// ListChains lists all the chains associated with a table
+	ListChains(table string) ([]string, error)
+	// ClearChain clears a chain in a table
+	ClearChain(table, chain string) error
+	// DeleteChain deletes a chain in the table. There should be no references to this chain
+	DeleteChain(table, chain string) error
+	// NewChain creates a new chain
+	NewChain(table, chain string) error
+	// Commit will commit changes if it is a batch provider.
+	Commit() error
+}
+
+// BatchProvider uses iptables-restore to program ACLs
+type BatchProvider struct {
+	ipt *iptables.IPTables
+
+	//        TABLE      CHAIN    RULES
+	rules       map[string]map[string][]string
+	batchTables map[string]bool
+
+	sync.Mutex
+}
+
+const (
+	restoreCmd = "iptables-restore"
+)
+
+// NewGoIPTablesProvider returns an IptablesProvider interface based on the go-iptables
+// external package.
+func NewGoIPTablesProvider(batchTables []string) (*BatchProvider, error) {
+	ipt, err := iptables.New()
+	if err != nil {
+		return nil, err
+	}
+	batchTablesMap := map[string]bool{}
+	for _, t := range batchTables {
+		batchTablesMap[t] = true
+	}
+
+	return &BatchProvider{
+		ipt:         ipt,
+		rules:       map[string]map[string][]string{},
+		batchTables: batchTablesMap,
+	}, nil
+}
+
+// Append will append the provided rule to the local cache or call
+// directly the iptables command depending on the table.
+func (b *BatchProvider) Append(table, chain string, rulespec ...string) error {
+	b.Lock()
+	defer b.Unlock()
+
+	if _, ok := b.batchTables[table]; !ok {
+		return b.ipt.Append(table, chain, rulespec...)
```

Comment:
Is this because for tables other than mangle, (nat etc) we use old method of programming iptables?

Reply:
yes. it's selective